File: spikes/cs/Server.Spike.NodeAccess/NodeManager.VariableNode.cs
// Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

namespace NodeAccess
{
    using Opc.UaFx;

    internal partial class NodeManager
    {
        private class VariableNode<T> : OpcDataVariableNode<T>
        {
            #region ---------- Private readonly fields ----------

            private readonly NodeManager manager;

            #endregion

            #region ---------- Public constructors ----------

            public VariableNode(NodeManager manager, IOpcNode parent, OpcName name, T value)
                : base(parent, name, value)
            {
                this.manager = manager;
            }

            #endregion

            #region ---------- Protected methods ----------

            protected override OpcAttributeValue<TAttribute> ReadAttributeValueCore<TAttribute>(
                    OpcReadAttributeValueContext context,
                    OpcAttributeValue<TAttribute> value)
            {
                var attribute = value.Attribute;

                // By default a user can not longer write to any node/attribute as soon the
                // OpcRequestType.Write is already denied (see Program.cs).
                //
                // To additionally inform the client/user that its access is restricted we shall
                // provide the according user dependent metadata for the UserAccessLevel and
                // UserWriteAccess attributes as well.
                if (attribute == OpcAttribute.UserAccessLevel || attribute == OpcAttribute.UserWriteAccess) {
                    if (!this.manager.CanWrite(context.Identity)) {
                        if (value is OpcAttributeValue<OpcAccessLevel> accessLevel) {
                            accessLevel = new OpcAttributeValue<OpcAccessLevel>(
                                    value.Attribute, accessLevel.Value & ~OpcAccessLevel.CurrentWrite);

                            return (OpcAttributeValue<TAttribute>)(object)accessLevel;
                        }

                        if (value is OpcAttributeValue<OpcAttributeWriteAccess> writeAccess) {
                            writeAccess = new OpcAttributeValue<OpcAttributeWriteAccess>(
                                    value.Attribute, OpcAttributeWriteAccess.None);

                            return (OpcAttributeValue<TAttribute>)(object)writeAccess;
                        }
                    }
                }

                return base.ReadAttributeValueCore(context, value);
            }

            #endregion
        }
    }
}

File: vb/Scenarios/Server.NodeAccess/NodeManager.VariableNode.vb
' Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

Imports Opc.UaFx

Namespace NodeAccess
    Partial Friend Class NodeManager
        Private Class VariableNode(Of T)
            Inherits OpcDataVariableNode(Of T)

            ' ---------- Private readonly fields ----------

            Private ReadOnly manager As NodeManager

            ' ---------- Public constructors ----------

            Public Sub New(ByVal manager As NodeManager, ByVal parent As IOpcNode, ByVal name As OpcName, ByVal value As T)
                MyBase.New(parent, name, value)
                Me.manager = manager
            End Sub

            ' ---------- Protected methods ----------

            Protected Overrides Function ReadAttributeValueCore(Of TAttribute)(
                    ByVal context As OpcReadAttributeValueContext,
                    ByVal value As OpcAttributeValue(Of TAttribute)) As OpcAttributeValue(Of TAttribute)
                Dim attribute = value.Attribute

                Dim accessLevel As OpcAttributeValue(Of OpcAccessLevel) = Nothing
                Dim writeAccess As OpcAttributeValue(Of OpcAttributeWriteAccess) = Nothing

                If (attribute = OpcAttribute.UserAccessLevel OrElse attribute = OpcAttribute.UserWriteAccess) Then
                    If (Not Me.manager.CanWrite(context.Identity)) Then
                        accessLevel = TryCast(CObj(value), OpcAttributeValue(Of OpcAccessLevel))

                        If (accessLevel IsNot Nothing) Then
                            accessLevel = New OpcAttributeValue(Of OpcAccessLevel)(
                                    value.Attribute,
                                    accessLevel.Value And Not OpcAccessLevel.CurrentWrite)

                            Return CType(CObj(accessLevel), OpcAttributeValue(Of TAttribute))
                        End If

                        writeAccess = TryCast(CObj(value), OpcAttributeValue(Of OpcAttributeWriteAccess))

                        If (writeAccess IsNot Nothing) Then
                            writeAccess = New OpcAttributeValue(Of OpcAttributeWriteAccess)(
                                    value.Attribute,
                                    OpcAttributeWriteAccess.None)

                            Return CType(CObj(writeAccess), OpcAttributeValue(Of TAttribute))
                        End If
                    End If
                End If

                Return MyBase.ReadAttributeValueCore(context, value)
            End Function
        End Class
    End Class
End Namespace
